Purge fix

cbosdo opened the pull request:
### What does this PR do?

* Adds a `type` attribute to each disk output by `virt.get_disks`
* Adds a `removables=True` parameter to `virt.purge`

### What issues does this PR fix or reference?

Fixes issue #46747 

### Previous Behavior

`virt.purge` removes all disk images, including ISOs and devices attached to removable devices (cdrom or floppy types)

### New Behavior

Now files attached to removable devices can be kept by calling `virt.purge removables=False`

### Tests written?

Yes

### Commits signed with GPG?

Yes

cbosdo committed:
virt.get_disks outputs the disk type

VM disks may also be floppies or cdroms. Output the libvirt device type
as a 'type' attribute of the disk for consumers to spot removable
devices (issue #46747).

cbosdo committed:
Add parameter to virt.purge to keep removable images.

Add a removables parameter to virt.purge to allow removing a VM without
removing the attached (and potentially shared) ISO removable images
(issue #46747).

cbosdo commented on salt/modules/virt.py:
```diff
@@ -1696,17 +1697,20 @@ def undefine(vm_):
     also delete the directories containing the virtual machine disk images -
     USE WITH EXTREME CAUTION!
 
+    Pass removables=False to avoid deleting cdrom and floppy images. NOTE THAT
```
@cachedout as mentioned by @gtmanfred in the issue: to avoid breaking the current behavior too hard.

gtmanfred replied:
If that is what we are doing, then it should be set to None, and then handled below, and also marked as deprecated behavior using warn_until.

https://docs.saltstack.com/en/latest/topics/development/deprecations.html

cbosdo replied:
done

gtmanfred commented on salt/modules/virt.py:
```diff
@@ -1696,19 +1697,29 @@ def undefine(vm_):
+            'The \'removables\' argument default value has been deprecated '
+            'and changed to False. Note that removable images will no longer '
+            'be purged unless explicitely setting it to True.'
+        )
```
you should also add removeables=True here.  We still want it to maintain the current behavior until that behavior is deprecated in Sodium.  Once we reach the time to work on the Sodium release, we will just remove this whole block, and set the default value in the definition to be `removables=False,`

And the message should be something like `The 'removables argument default value is True, but will be changed to False by default in {version}.  Please set to True to maintain the current behavior in the future.`

rallytime reviewed (changes requested):
I have one small request here, but otherwise, this looks good.

rallytime commented:
re-run centos

rallytime commented:
@cbosdo There are some related unit tests failing due to this change:

- unit.modules.test_virt.VirtTestCase.test_get_disks
- unit.modules.test_virt.VirtTestCase.test_purge_default
- unit.modules.test_virt.VirtTestCase.test_purge_noremovable

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-cent7-py3/3907/

Can you take a look?

cbosdo commented:
@rallytime those tests are all failing with message `No such file or directory: 'qemu-img'` which means that some package seems to be missing on the setup since my commits didn't change anything related to qemu-img.

rallytime commented:
@cbosdo Ah, then something in your changes seems to be triggering those tests when they weren't before, though I don't know what it is off hand. We need to be protecting those tests against running when that library is missing. Would you mind adding that check?

cbosdo commented:
@rallytime my bad, I added these tests and get_disks is using qemu-img. I fixed that my mocking up subprocess.Popen and subprocess.Popen.communicate to fake qemu-img's call. Should be back to normal now

rallytime commented:
Thanks @cbosdo!
